li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaInstrumentedProducerImpl.java

```diff
@@ -312,11 +312,13 @@ public void abortTransaction() throws ProducerFencedException {
   public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
     verifyOpen();
 
-    delegateLock.readLock().lock();
+    //the callback may try and obtain a write lock (say call producer.close())
+    //so we grab an update lock, which is upgradable to a write lock
+    delegateLock.updateLock().lock();
```

Comment:
The java doc said it will block other updatelock :
https://github.com/npgall/concurrent-locks
// allows other readers, blocks others from acquiring update or write locks

Does it prevent other senders since you tried to grab update lock instead of read lock.

Reply:
very good catch. thank you